### PR TITLE
feat: add submodule with default base access group set

### PR DESCRIPTION
**Pull request by MatthewLemmond**
### Description
Add the deployable architecture code to provision a default set of access groups 

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: .catalog-onboard-pipeline.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  - name: deploy-arch-ibm-access-management
+    kind: solution
+    catalog_id: 7df1e4ca-d54c-4fd0-82ce-3d13247308cd
+    offering_id: 48d2a13c-f353-47cb-a9c4-6fb560190ccc
+    variations:
+      - name: standard
+        mark_ready: false
+        install_type: fullstack
+        scc:
+          # Instance ID in account "GoldenEye FSCloud Testing (ae14871c179f49869e19657b37ec7c4e) <-> 2626044
```
Comment (ocofaigh) [inline .catalog-onboard-pipeline.yaml:13]: This repo will us the GEdev account, so remove this comment and set instance id to `1c7d5f78-9262-44c3-b779-b28fe4d88c37`

File: ibm_catalog.json [hunk 2]
```diff
@@ -0,0 +1,87 @@
+{
+  "products": [
+    {
+      "name": "deploy-arch-ibm-access-group",
```
Comment (ocofaigh) [inline ibm_catalog.json:4]: This needs to be `deploy-arch-ibm-access-management` as thats the name I submitted for approval

File: ibm_catalog.json [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      "tags": [
+        "ibm_created",
+        "target_terraform",
+        "terraform",
+        "devops",
+        "security",
+        "solution"
+      ],
+      "keywords": [
+        "base layer",
```
Comment (ocofaigh) [inline ibm_catalog.json:16]: remove

File: ibm_catalog.json [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        "IaC",
+        "infrastructure as code",
+        "terraform",
+        "solution",
+        "account",
```
Comment (ocofaigh) [inline ibm_catalog.json:21]: maybe replace with access? access-group? roles?

File: ibm_catalog.json [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        "iam"
+      ],
+      "short_description": "Creates and configures the base layer access groups of an IBM Cloud account",
+      "long_description": "This architecture supports creating and configuring the access groups of an account, includes a default set of access groups for Administrators, Privileged Users, and Observers broken up by category of service.",
+      "offering_docs_url": "https://github.com/terraform-ibm-modules/terraform-ibm-iam-access-group/blob/main/README.md",
```
Comment (ocofaigh) [inline ibm_catalog.json:26]: point to the DA directory readme

File: solutions/access-management/version.tf [hunk 6]
```diff
@@ -0,0 +1,10 @@
+terraform {
+  required_version = ">= 1.3.0, <1.7.0"
+  required_providers {
+    # Use "greater than or equal to" range in modules
```
Comment (ocofaigh) [inline solutions/access-management/version.tf:4]: update the code comment to say: `# Lock DA into an exact provider version - renovate automation will keep it updated`

And also add the rule to renovate.json file (same as [this one](https://github.com/terraform-ibm-modules/terraform-ibm-scc-da/blob/main/renovate.json))

File: solutions/access-management/version.tf [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+    ibm = {
+      source  = "ibm-cloud/ibm"
+      version = "1.51.0"
```
Comment (ocofaigh) [inline solutions/access-management/version.tf:7]: use 1.53.0 (the current latest)

File: tests/pr_test.go [hunk 8]
```diff
@@ -47,3 +49,50 @@ func TestRunBasicUpgradeExample(t *testing.T) {
...
+	t.Parallel()
+
+	options := setupDAOptions(t, "ag-solution", daDir)
+
+	output, err := options.RunTestConsistency()
+	assert.Nil(t, err, "This should not have errored")
+	assert.NotNil(t, output, "Expected some output")
+}
+
+func TestRunDAUpgrade(t *testing.T) {
```
Comment (ocofaigh) [inline tests/pr_test.go:88]: I guess you can now remove `TestRunBasicUpgradeExample` - we dont need 2 upgrade tests

File: solutions/access-management/variables.tf [hunk 9]
```diff
@@ -0,0 +1,1553 @@
...
+  description = "A map of policies for the Administrator Observability access group, has a set of default policies that can be overridden"
+  default = {
+    "atracker" = {
+      roles = ["Administrator", "Writer"]
+      resources = [{
+        service = "atracker"
+      }]
+      tags = []
+    }
+    "observability" = {
```
Comment (ocofaigh) [inline solutions/access-management/variables.tf:46]: observability is logging and monitoring, however you have logdna below already, so should this be renamed monitoring?

File: solutions/access-management/variables.tf [hunk 10]
```diff
@@ -0,0 +1,1553 @@
...
+      roles = ["Administrator", "Manager"]
+      resources = [{
+        service = "sysdig-monitor"
+      }]
+      tags = []
+    }
+    "logdna" = {
```
Comment (ocofaigh) [inline solutions/access-management/variables.tf:53]: use logging instead of logdna

File: solutions/access-management/variables.tf [hunk 11]
```diff
@@ -0,0 +1,1553 @@
...
+
+variable "admin_observability_ag_description" {
+  type        = string
+  description = "Description of the admin observability access group"
+  default     = null
+}
+
+variable "admin_observability_ag_tags" {
+  type        = list(string)
+  description = "Tags that should be applied to the service"
```
Comment (ocofaigh) [inline solutions/access-management/variables.tf:91]: service ?

File: solutions/access-management/variables.tf [hunk 12]
```diff
@@ -0,0 +1,1553 @@
...
+
+variable "admin_security_ag_description" {
+  type        = string
+  description = "Description of the admin security access group"
+  default     = null
+}
+
+variable "admin_security_ag_tags" {
+  type        = list(string)
+  description = "Tags that should be applied to the service"
```
Comment (ocofaigh) [inline solutions/access-management/variables.tf:198]: service?

**Review by ocofaigh — CHANGES_REQUESTED** [review]
I left some comments. Also requesting review from @vburckhardt 

PS: Can you add the `deployable-architecture` topic to `.github/settings.yml`?

File: solutions/access-management/variables.tf (hunk 11 above)
Comment (MatthewLemmond) [reply solutions/access-management/variables.tf:91]: Description taken from https://github.com/terraform-ibm-modules/terraform-ibm-iam-access-group/blob/main/variables.tf#L34, unsure on what it should be changed to

File: solutions/access-management/variables.tf (hunk 12 above)
Comment (MatthewLemmond) [reply solutions/access-management/variables.tf:198]: Description taken from https://github.com/terraform-ibm-modules/terraform-ibm-iam-access-group/blob/main/variables.tf#L34, unsure on what it should be changed to

**Review by ocofaigh — APPROVED** [review]
Might need another round of reviews, but approving for now so we can get this into the catalog. 
@vburckhardt do you wan't to do a final review before I merge?

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@MatthewLemmond After some talks today, we decided that instead of this being a DA, it should be its own module. Can you move this into a `modules` folder? You may need to rework a few things, but hopefully straight forward. They key being of course it now wont have a provider.tf in it

**Comment by MatthewLemmond** [comment]
Per conversation with @ocofaigh this PR has been moved from a solution/DA to a submodule that will be called in other DAs

File: modules/access-management/README.md [hunk 13]
```diff
@@ -0,0 +1,12 @@
+# IBM Cloud Access Management module
+
```
Comment (ocofaigh) [inline modules/access-management/README.md:2]: This readme needs to be updated with a usage, and the terradocs metadata so the terradocs hooks generates inputs, outputs etc

File: modules/access-management/main.tf [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+  policies          = each.value.policies
+  dynamic_rules     = each.value.dynamic_rules
+  add_members       = each.value.add_members
+  description       = each.value.description
+  tags              = each.value.tags
+  ibm_ids           = each.value.ibm_ids
+  service_ids       = each.value.service_ids
+}
+
+resource "ibm_iam_custom_role" "custom_iam_roles" {
```
Comment (ocofaigh) [inline modules/access-management/main.tf:140]: I think I raised this before, but what if one of these custom roles need to be included in one of the access groups?

**Review by ocofaigh — CHANGES_REQUESTED** [review]
some comments

File: modules/access-management/main.tf (hunk 14 above)
Comment (ocofaigh) [reply modules/access-management/main.tf:140]: as per discussions we can remove custom role support for first release of submodule

Comment (MatthewLemmond) [reply modules/access-management/main.tf]: Removed custom roles from the submodule

File: cra-config.yaml [hunk 15]
```diff
@@ -1,11 +1,6 @@
 # More info about this file at https://github.com/terraform-ibm-modules/common-pipeline-assets/blob/main/.github/workflows/terraform-test-pipeline.md#cra-config-yaml
 version: "v1"
 CRA_TARGETS:
-  - CRA_TARGET: "examples/basic" # Target directory for CRA scan. If not provided, the CRA Scan will not be run.
+  - CRA_TARGET: "modules/access-management" # Target directory for CRA scan. If not provided, the CRA Scan will not be run.
```
Comment (ocofaigh) [inline cra-config.yaml:4]: CRA needs to be run against an example / solution that has a provider.tf in it. Its not supposed to run against a module

File: examples/access-management/version.tf [hunk 16]
```diff
@@ -0,0 +1,10 @@
+terraform {
+  required_version = ">= 1.3.0, <1.7.0"
+  required_providers {
+    # Lock DA into an exact provider version - renovate automation will keep it updated
```
Comment (ocofaigh) [inline examples/access-management/version.tf:4]: remove this comment, its not a DA

File: modules/access-management/version.tf [hunk 17]
```diff
@@ -0,0 +1,11 @@
+terraform {
+  required_version = ">= 1.3.0, <1.7.0"
+  required_providers {
+    # Provider used by the module, ignoring linter warning
+    # tflint-ignore: terraform_unused_required_providers
+    ibm = {
+      source  = "ibm-cloud/ibm"
+      version = ">= 1.51.0"
```
Comment (ocofaigh) [inline modules/access-management/version.tf:8]: `>= 1.51.0, <2.0.0`

**Review by ocofaigh — CHANGES_REQUESTED** [review]
left some minor comments - @vburckhardt would you mind reviewing to ensure this is what you had in mind for the module?

File: cra-config.yaml (hunk 15 above)
Comment (MatthewLemmond) [reply cra-config.yaml:4]: updated to `examples/access-management`

**Comment by ocofaigh** [comment]
Pipeline issue should be resolved now @MatthewLemmond if you rebase and try again

File: modules/access-management/variables.tf [hunk 18]
```diff
@@ -0,0 +1,1535 @@
...
+
+variable "admin_observability_ag_description" {
+  type        = string
+  description = "Description of the admin observability access group"
+  default     = null
+}
+
+variable "admin_observability_ag_tags" {
+  type        = list(string)
+  description = "Tags that should be applied to the service"
```
Comment (jor2) [inline modules/access-management/variables.tf:85]: `The list of tags that you want to associated with your admin observability access group` maybe? same for the others

File: modules/access-management/variables.tf [hunk 19]
```diff
@@ -0,0 +1,1535 @@
...
+
+variable "admin_network_ag_description" {
+  type        = string
+  description = "Description of the admin network access group"
+  default     = null
+}
+
+variable "admin_network_ag_tags" {
+  type        = list(string)
+  description = "Tags that should be applied to the service"
```
Comment (jor2) [inline modules/access-management/variables.tf:355]: `The list of tags that you want to associated with your admin network access group`

File: modules/access-management/variables.tf [hunk 20]
```diff
@@ -0,0 +1,1535 @@
...
+
+variable "privileged_observability_ag_description" {
+  type        = string
+  description = "Description of the privileged observability access group"
+  default     = null
+}
+
+variable "privileged_observability_ag_tags" {
+  type        = list(string)
+  description = "Tags that should be applied to the service"
```
Comment (jor2) [inline modules/access-management/variables.tf:597]: `The list of tags that you want to associated with your privileged observability access group`

**Review by jor2 — COMMENTED** [review]
small comments

**Review by jor2 — APPROVED** [review]

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**